lib/eco/indentmanager.py

```diff
@@ -196,32 +189,38 @@ def apply_nodes(self, new, bol):
                 new.append(self.create_token("dedent"))
             new.append(self.create_token("newline"))
 
-            changed = self.indentation_nodes_changed(node, new)
-            if changed:
-                self.changed = True
-                # remove old indentation nodes
-                self.remove_indentation_nodes(node)
-                # and insert new ones
-                for n in new:
-                    node.insert_after(n)
-
-    def indentation_nodes_changed(self, bol, nodes):
-        """Compares indentation tokens in a line with new tokens to find out
-        whether the line needs to be updated"""
+            self.merge_nodes(node, new)
+
+    def merge_nodes(self, bol, newnodes):
         previous_nodes = []
+        last_node = bol
         node = bol.next_term
         while isinstance(node.symbol, IndentationTerminal):
             if not node.deleted:
                 previous_nodes.append(node)
             node = node.next_term
 
-        previous_nodes.reverse()
-        if len(previous_nodes) != len(nodes):
-            return True
-        for i in range(len(nodes)):
-            if nodes[i].symbol != previous_nodes[i].symbol:
-                return True
-        return False
+        for i in range(len(previous_nodes)):
+            try:
+                newnode = newnodes.pop()
+            except IndexError:
+                # remove leftover previous_nodes
+                previous_nodes[i].remove()
+                continue
+
+            oldnode = previous_nodes[i]
+            if newnode.symbol != oldnode.symbol:
+                # remove old, insert new
+                oldnode.replace(newnode)
+                last_node = newnode
+            else:
+                if oldnode.deleted:
+                    oldnode.deleted = False
+                last_node = oldnode
+
+        for n in newnodes:
+            # insert remaining newnodes
```

Comment:
It's kinda part of `remove leftover previous nodes` and `remove old, insert new`. I don't think it adds less than those comments.